pkg/pillar/containerd/containerd.go

```diff
@@ -246,6 +250,317 @@ func (client *Client) CtrDeleteBlob(ctx context.Context, blobHash string) error
 	return client.contentStore.Delete(ctx, digest.Digest(blobHash))
 }
 
+// CtrCreateImageFromRootfs creates an OCI image from an existing rootfs directory
+func (client *Client) CtrCreateImageFromRootfs(ctx context.Context, imageName, rootfsPath string) error {
+	if err := client.verifyCtr(ctx, true); err != nil {
+		return fmt.Errorf("CtrCreateImageFromRootfs: exception while verifying ctrd client: %s", err.Error())
+	}
+
+	// Create a tarball from the rootfs directory
+	tarPath := "/tmp/rootfs.tar"
```

Comment:
And /tmp/rootfs.tar will be overlayfs i.e., this will use up memory.

Reply:
I have been thinking about this. You need to know the digest when you first open the channel to push it to containerd, but you need to have the whole thing ready to get the hash. I was going to suggest to do it twice: first stream it to `/dev/null` while hashing, and get the digest, then stream it to containerd with the results of the first. But tar files are only somewhat reliably reproducible. I have done it, but do we want to depend on that? You probably could, if you control the dates tightly. If that is a path we want to follow, let me know, I can find the tar code where I did it.

It still bothers me: we built this xen-tools rootfs _from an existing image_ when we built the rootfs. Sure, it was laid out on the filesystem because it was useful there and not in its blobs. Now we are trying to recreate it, and doing all sorts of fun things with it? I kind of want us to be able to just embed it in the file to start with. Not actually this way, but the equivalent of:

1. `docker save xen-tools > /rootfs/images/xen-tools.tar`
2. In `pillar`, bind-mount `/rootfs/images:/images`
3. Then we can just use `/images/xen-tools.tar`

The issue is size. This now becomes a permanent part of the image, another 26MB. That seems a waste.